### ISSUE TITLE
Debugger out-of-memory when stepping

### ISSUE THREAD

**Issue by svaarala** [svaarala/duktape#1117]
When the debugger is paused, no garbage is freed to avoid disturbing any heap inspection related pointers being queried by the debug client. Garbage accumulated during paused state (e.g. Eval) is intended to be collected when execution resumes (Resume, StepXxx, etc).

However, if you just keep on stepping and evaling values as you step, it may be that a voluntary GC is not executed in time: one might then be paused, Eval something, run out of memory, and fail to run GC because execution is paused.

This was originally an intentional decision to ensure resume/step wouldn't always need to trigger mark-and-sweep (which can be a bit slow for large heaps). But this has the unintended side effect described above.

The easiest fix is just to force a GC at some useful moment, e.g.:
- When resuming (or stepping)
- When becoming paused
- Every time when debugger message loop is entered in paused state

This should be fixed in Duktape 1.5.x maintenance branch too because it impacts very low memory targets.

**Comment by svaarala** [comment]
Somewhat related: for low memory targets that don't do heap inspection, having a debugger command to force a GC during paused state would be useful. This was part of the original heap inspection plan but was not implemented at the time.

**Comment by svaarala** [comment]
For low memory targets the alternative design of keeping GC running in paused state and requiring an explicit GC lock when doing heap inspection might be cleaner. This would allow GC within an Eval, for example, which isn't possible with an explicit "force GC" command. So you can still run out of memory inside a single Eval.

**Comment by svaarala** [comment]
Notes on IRC discussion:
- Adding a "force GC" to Duktape 1.6.0 would be quite simple.
- Adding a config option to enable/disable the automatic GC freezing during paused state would be possible (incompatible with heap inspection though). Maybe allow GC in paused state if heap inspection is disabled (there's already a config option for that).

As for the high level alternatives for GC and heap inspection interaction:
- Keep current, add automatic GC on resume (or some other opportune place), and add a forced GC. Has the limitations described above (e.g. single Eval runs out of memory).
- Keep current, but make GC locking explicitly requested by debug client. Heap inspection commands would fail with an error message (e.g. "gc not locked") until debug client did that. Reset GC locking on resume. This makes a single Eval behave better as it can GC while doing the eval.

**Comment by svaarala** [comment]
@fatcerberus This issue (and its resolution) might interest you. I'd like to find a solution for 1.x, and if there's any change of model, for 2.0.0. This is one of the issues we discussed but decided not to do anything about it yet.

I would be fine with the explicit GC locking approach, provided that the dangerous commands (inspection) would error out until the debug client locked the GC.

Also adding a forced GC to resume/step paths would be a simple (but not ideal) fix for the basic stepping issue.

**Comment by fatcerberus** [comment]
Despite my misgivings when we originally discussed this, I think I would be okay in general with an explicit GC locking command.  The downside is a bit of extra latency for step-eval-step workflows as the client has to send an extra command each time, but sacrifices must be made. :)

Maybe to be safe, Duktape could reject any heap pointers sent by the client, regardless of context, with a GC-not-locked error.  This way all potential for unsafe operations is cut off at the pass.

**Comment by svaarala** [comment]
Regarding latency, there shouldn't be too much latency because the debug client doesn't need to actually wait for any responses. It can just e.g. send: Pause, LockGc, Eval, without waiting for responses before sending each command. Once it has the Eval result, it can do inspection; finally, when resuming it can send UnlockGc + Resume without waiting for results - or more likely, Resume would automatically unlock the GC. (There are some use cases for keeping GC locked through step/resume but not very many.)

Regarding safety, did you mean that the debug client would refuse to parse heap pointers when not locked? Or that relevant commands would do so? I was thinking more of the commands themselves but it could also be implemented in the value parsing too. There might be some future cases where that would be inappropriate (e.g. when just stringifying a pointer argument one way or another) but for now that should be OK.

**Comment by fatcerberus** [comment]
Right, that's what I meant: Duktape would refuse heap pointers on the *target* side until the GC is locked.  But I'm sure if there are any cases where it's both useful and safe to accept them without a lock in place.

**Comment by svaarala** [comment]
But do you mean the target would refuse to parse the heap pointer dvalues - or requests containing them?

**Comment by fatcerberus** [comment]
Mostly I'm just thinking of how to reduce code duplication to minimize pitfalls when adding commands in the future.  I guess for that it would be enough to have a "read heap pointer, reject if GC not locked" helper.

**Comment by fatcerberus** [comment]
By the way, for some reason all this time I thought Duktape _did_ run a mark-and-sweep when resuming, but it seems that wasn't the case after all... :)

**Comment by roceh** [comment]
Would any solution work for the situation where multiple evals are issued while paused? As currently the vscode debugger will issue quite a few and by default they take a few k ram each time to run. Also as you hover over variables they will each call an eval, so if there is no GC when paused eventually memory runs out.  If the eval wasn't so expensive it wouldn't be such an issue.

**Comment by svaarala** [comment]
@fatcerberus Re: running GC on resume, that was my recollection too - but I think we changed it not to do so.

**Comment by svaarala** [comment]
@roceh Adding a `ForceGc` command that the debug client can issue would help in that Eval scenario. The debug client would need to guarantee that it won't hold any Duktape heap pointers across that command though, because any such pointers would be potentially invalidated.

Also an explicit LockGc + UnlockGc command pair would work: Duktape would run with GC enabled even when paused. All Evals would be GC'd etc. The debug client would issue LockGc *only* when doing heap pointer inspection and can then UnlockGc when it is done - again enabling automatic GC.

**Comment by fatcerberus** [comment]
One important thing to stress in documentation is that, if one wants to inspect Eval results, GC must be locked BEFORE the Eval.  Which now brings us back to the issue of a single Eval running you out of memory...

**Comment by fatcerberus** [comment]
Hm, locking down the garbage collector entirely is a rather large hammer in any case.

Would it be possible to implement a mechanism where only the Eval result (and anything reachable from it) is excluded from collection?

**Comment by svaarala** [comment]
> Would it be possible to implement a mechanism where only the Eval result (and anything reachable from it) is excluded from collection?

The simplest way to do that would be to simply ensure it is reachable, e.g. keep a reference to the last Eval result somewhere (and lose the reference on resume). Anything else gets pretty complicated quite fast. But I'm not sure that would work going forward. For example, you might be inspecting weakly referenced values which are liable to be collected without notice even if the Eval result itself was strongly referenced.

For Eval it would also be possible to first run the Eval with GC enabled. When Eval is done, before losing the result reference, lock the GC. Once that's done, respond to the debug client and pop the result. This way GC would be used normally during eval but the GC would be frozen at the last minute without the Eval result itself being collected.

The downside of that is that GC locking and individual commands would have hard-to-understand interactions. You might also need a flag for Eval to "lock after Eval is done", etc.

**Comment by fatcerberus** [comment]
Here's a more general-purpose idea that would also accomodate, e.g. debugger watch lists: Add some kind of "debugger stash" and commands for stashing/unstashing heap pointers.  Eval could then take a flag that says whether or not to automatically stash the result.  That's needed because in the absence of a global GC lock, it must be atomic.

**Comment by svaarala** [comment]
Using a stash (or "scratch") based approach would also work with a debugger command set centered around the stash: most operations could write results to stash slots requested by the debug client, and there would be commands for inspecting stash slots, clearing them, etc. If the debug client didn't care about storing the results for longer term, it could just reuse the same slot for all Eval results for example.

**Comment by svaarala** [comment]
@fatcerberus And that reminds me actually of one more approach that was on the table when the debugger was initially designed: allow the debug client to interact with the *value stack* when it is executing commands. I think we discussed this also when implementing application specific methods.

When paused, an Eval command would leave its result on the target's value stack. You would then have the ability to e.g. inspect the value stack at index -1 (the eval result), pop it off, etc.

This would have the upside of the value stack acting as a logical "scratch" and it would work the same way as the API works for C code. For example, if the debugger had a command for "protected call function", you could push the function and its arguments using whatever commands are best suited to your situation and invoke a ProtectedCall. For example, the function might be a setter you obtained from inspection, while the setter argument could be pushed using an Eval command.

It would also be logical and clear that the value stack would be cleared when you step or resume. The debug client would ensure you don't look beyond the initial stack, i.e. the stack top when you became paused, or at least pop off too many values.

**Comment by svaarala** [comment]
For practical code changes there are two separate issues here:
- Incremental improvement for 2.0 and 1.6 - something useful enough, but requiring only small changes to Duktape and debug clients.
- Rethinking the inspection model for 3.0.

**Comment by fatcerberus** [comment]
The value stack approach has its merits, however I think we would still need some sort of persistent stash that wouldn't be cleared automatically on resume.  For example, as alluded to above, a client might set up a watch list, and would need the watched objects' heap pointers to remain valid as long as they remain in the watch list.

**Comment by svaarala** [comment]
Some of the pros/cons ultimately depend on what kind of debug client use cases are supported. For example, some debug clients might consist of relatively independent modules updating their own "panels" without much communication between them (this is what I would like to do, for example). The value stack approach doesn't work very well for this because the value stack would be shared - and since the commands are obviously not atomic w.r.t. the value stack, it requires client side coordination. For other kinds of clients this is not a limitation at all.

The most promising approaches at this point are IMO:
- Some form of explicit or automatic locking which is good enough. This includes possibilities like locking after Eval is complete but before the result is GC'd.
- Some kind of stash approach.

**Comment by fatcerberus** [comment]
Locking doesn't work for the watch list use case though unless you keep GC locked for the duration of runtime, which would run you out of memory very fast :)

**Comment by svaarala** [comment]
For a debugger stash one easy and intuitive approach is that a debugger stash is created when one becomes paused. The stash would then be GC'd when we resume (simply by setting the stash object to null and letting refcount do its job).

Adding separate support for persistent stashing (probably limited by debugger attached/detached session lifetime?) would be easy by adding a separate stash for that, tied to the session. But it would have impact on the debugger command syntax (to interact with both stashes) and wouldn't be the simplest model for the debug client.

Another approach is that there's a single stash but when one resumes there's some way to know which references need to be wiped. Maybe property attributes could be used for that purpose somehow. The debugger would require a command to mark a certain reference persistent.

GC pinning is not really a solution, because one needs book-keeping state to undo the pinning when e.g. a debugger session lifecycle ends. It's possible to implement it so that there would be an explicit "pinned by debugger" heap object flag so that it'd be possible to walk the heap and clear all the flags on a resume; this would be more lightweight than an actual stash. But somewhat awkward overall and inflexible if the pinning lifetime would need more flexibility later.

**Comment by svaarala** [comment]
> Locking doesn't work for the watch list use case though unless you keep GC locked for the duration of runtime, which would run you out of memory very fast :)

That's true - for the short term change (which this issue is really for) that's not a concern because that use case is not supported at all.

But I agree it'd be nice to solve in the longer term. Only some form of fine grained pinning / persisted reference would solve that.

**Comment by fatcerberus** [comment]
Yeah, sorry, when issues like this come up, I tend to very easily lapse into thinking in the general case.  I have to get my head out of the clouds now ;)

**Comment by svaarala** [comment]
Hehe, I don't mind at all. I have the same tendency ;)

Like I said above there are two issues here: one for the shorter term, and another for the longer. Heap object inspection (and the debugger protocol as a whole) is a "version 1" with some successful design choices and some less successful.

I don't mind fixing the less successful choices, but when that involves changing the basic structure it's good to be really sure the change is an overall improvement (and hopefully a simplification). Opened #1118 for that.
